man/ngircd.8
.\"
.\" $Id: ngircd.8,v 1.11 2005/08/12 13:20:54 alex Exp $
.\"
.TH ngircd 8 "August 2005" ngircd "ngIRCd Manual"
.SH NAME
ngIRCd \- the next generation IRC daemon
.SH SYNOPSIS
.B ngircd [
.I Options
.B ]
.SH DESCRIPTION
.BR ngIRCd
is a free open source daemon for the Internet Relay Chat (IRC),
developed under the GNU General Public License (GPL).
.PP
It's written from scratch and is not based upon the original IRCd like
many others. It is easy to configure, supports server links (even with
original ircd's) and runs on hosts with changing IP addresses (such as
dial-in networks).
.PP
Currently supported platforms include AIX, A/UX, FreeBSD, HP-UX, IRIX,
Linux, Mac OS X, NetBSD, OpenBSD, Solaris, and Windows with Cygwin.
.PP
As ngIRCd relies on UNIX standards and uses GNU automake and GNU autoconf
there are good chances that it also supports other UNIX-based operating
systems as well.
.SH OPTIONS
The default behaviour of
.BR ngircd
is to read its standard configuration file (see below), to detach from the
controlling terminal and to wait for clients.
.PP
You can use these options to modify this default:
.TP
\fB\-f\fR \fIfile\fR, \fB\-\-config\fR \fIfile\fR
Use
.I file
as configuration file.
.TP
\fB\-n\fR, \fB\-\-nodaemon\fR
Don't fork a child and don't detach from controlling terminal.
All log messages go to the console and you can use CTRL-C to
terminate the server.
.TP
\fB\-p\fR, \fB\-\-passive\fR
Disable automatic connections to other servers. You can use the IRC command
CONNECT later on as IRC Operator to link this ngIRCd to other servers.
.TP
\fB\-t\fR, \fB\-\-configtest\fR
Read, validate and display the configuration; then exit.
.TP
\fB\-\-version\fR
Output version information and exit.
.TP
\fB\-\-help\fR
Display a brief help text and exit.
.SH FILES
.I /usr/local/etc/ngircd.conf
.RS
The system wide default configuration file.
.RE
.I /usr/local/etc/ngircd.motd
.RS
Default "message of the day" (MOTD).
.RE
.SH AUTHOR
Alexander Barton,
.UR mailto:alex@barton.de
alex@barton.de
.UE
.br
Homepage:
.UR http://ngircd.barton.de/
http://ngircd.barton.de/
.UE
.SH "SEE ALSO"
.BR ngircd.conf (5),
.BR ircd (8)
.\"
.\" -eof-


NEWS

                     ngIRCd - Next Generation IRC Server

                        (c)2001-2005 Alexander Barton,
                    alex@barton.de, http://www.barton.de/

               ngIRCd is free software and published under the
                   terms of the GNU General Public License.

                                  -- NEWS --


ngIRCd 0.9.0 (2005-07-24)

  - Never run with root privileges but always switch the user ID.
  - Make "netsplit" messages RFC compliant.
  - Implemented the IRC function "WHOWAS".
  - New configuration option "OperServerMode" to enable a workaround needed
    when running an network with ircd2 servers and "OperCanUseMode" enabled
    to prevent the ircd2 daemon to drop mode changes of IRC operators.
    Patch by Florian Westphal, <westphal@foo.fh-furtwangen.de>.
  - Implemented support for "secret channels" (channel mode "s").
  - New configuration option "Mask" for [Operator] sections to limit OPER
    commands to users with a specific IRC mask. Patch from Florian Westphal.
  - New configuration variable "PidFile", section "[Global]": if defined,
    the server writes its process ID (PID) to this file. Default: off.
    Idea of Florian Westphal, <westphal@foo.fh-furtwangen.de>.
  - Added support for the Howl (http://www.porchdogsoft.com/products/howl/)
    Rendezvous API, in addition to the API of Apple (Mac OS X). The available
    API will be autodetected when you call "./configure --with-rendezvous".

ngIRCd 0.8.0 (2004-06-26)

  - Two new configuration options: "ChrootDir" and "MotdPhrase", thanks to
    Benjamin Pineau <ben@zouh.org>. Now you can force the daemon to change
    its root and working directory to something "safe". MotdPhrase is used
    to define an "MOTD string" instead of a whole file, useful if the
    "real" MOTD file would be outside the "jail".
  - INVITE- and BAN-lists become synchronized between IRC+ servers when
    establishing new connections, if the peer supports this as well.
  - The type of service (TOS) of all sockets is set to "interactive" now.
  - Added short command line option "-t" as alternative to "--configtest".
  - Added optional support for "IDENT" lookups on incoming connections. You
    have to enable this function with the ./configure switch "--with-ident".
    The default is not to do IDENT lookups.

ngIRCd 0.7.5 (2003-07-11)

  - New configuration variable "MaxConnectionsIP" to limit the number of
    simultaneous connections from a single IP that the server will accept.
    This configuration options lowers the risk of denial of service attacks
    (DoS), the default is 5 connections per client IP.
  - Added new configuration variable "Listen" to bind all listening
    sockets of the server to a single IP address.
       
ngIRCd 0.7.1 (2003-07-18)

  - Added support for GNU/Hurd.

ngIRCd 0.7.0 (2003-05-01)

  - New command CONNECT to enable and add server links. The syntax is not
    RFC-compatible: use "CONNECT <name> <port>" to enable and connect an
    configured server and "CONNECT <name> <port> <host> <mypwd> <peerpwd>"
    to add a new server (ngIRCd tries to connect new servers only once!).
  - Added DISCONNECT command ("DISCONNECT <name>") to disable servers.
  - New command TRACE (you can trace only servers at the moment).
  - New command HELP that lists all understood commands.
  - ngIRCd can register itself with Rendezvous: to enable support pass the
    new switch "--with-rendezvous" to configure.
  - Added support for TCP Wrappers library: pass "--with-tcp-wrappers" to
    configure to enable it.
  - Changed some configure options to use "--with"/"--without" as prefix
    instead of "--enable"/"--disable": "--without-syslog", "--without-zlib",
    "--with-tcp-wrappers", and "--with-rendezvous".
  - Enhanced manual pages ngircd(8) and ngircd.conf(5).
  - Documentation is now installed in $(datadir)/doc/ngircd.


Older news (sorry, only available in german language):

ngIRCd 0.6.0, 2002-12-24

  - beim Schliessen einer Verbindung zeigt der Server nun vor dem ERROR
    noch eine Statistik ueber die empfangene und gesendete Datenmenge an.
  - Connection-Strukturen werden nun "pool-weise" verwaltet; der Pool wird
    bei Bedarf bis zu einem konfigurierten Limit vergroessert.
  - Mit der neuen Konfigurationsvariable "MaxConnections" (Sekion "Global")
    kann die maximale Anzahl gleichzeitiger Verbindungen begrenzt werden.
    Der Default ist -1, "unlimitiert".
  - der Server erkennt nun, ob bereits eine eingehende Verbindung von einem
    Peer-Server besteht und versucht dann nicht mehr, selber eine eigene
    ausgehende Verbindung zu diesem auufzubauen. Dadurch kann nun auf beiden
    Servern in der Konfiguration ein Port fuer den Connect konfiguriert
    werden (beide Server versuchen sich dann gegenseitig zu connectieren).
  - Server identifizieren sich nun mit asyncronen Passwoertern, d.h. das
    Passwort, welches A an B schickt, kann ein anderes sein als das, welches
    B als Antwort an A sendet. In der Konfig.-Datei, Abschnitt "Server",
    wurde "Password" dazu durch "MyPassword" und "PeerPassword" ersetzt.

src/ngircd/ngircd.c
/*
 * ngIRCd -- The Next Generation IRC Daemon
 * Copyright (c)2001-2005 by Alexander Barton (alex@barton.de)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * Please read the file COPYING, README and AUTHORS for more information.
 */


#include "portab.h"

static char UNUSED id[] = "$Id: ngircd.c,v 1.108 2005/07/31 20:13:08 alex Exp $";

/**
 * @file
 * The main program, including the C function main() which is called
 * by the loader of the operating system.
 */

#include "imp.h"
#include <assert.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <string.h>
#include <unistd.h>
#include <time.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <fcntl.h>
#include <pwd.h>
#include <grp.h>

#include "defines.h"
#include "resolve.h"
#include "conn.h"
#include "client.h"
#include "channel.h"
#include "conf.h"
#include "cvs-version.h"
#include "lists.h"
#include "log.h"
#include "parse.h"
#include "irc.h"

#ifdef ZEROCONF
#include "rendezvous.h"
#endif

#include "exp.h"
#include "ngircd.h"


static void Initialize_Signal_Handler PARAMS(( void ));
static void Signal_Handler PARAMS(( int Signal ));

static void Show_Version PARAMS(( void ));
static void Show_Help PARAMS(( void ));

static void Pidfile_Create PARAMS(( long ));
static void Pidfile_Delete PARAMS(( void ));

static void Fill_Version PARAMS(( void ));

static void Setup_FDStreams PARAMS(( void ));

static bool NGIRCd_Init PARAMS(( bool ));

/**
 * The main() function of ngIRCd.
 * Here all starts: this function is called by the operating system loader,
 * it is the first portion of code executed of ngIRCd.
 * @param argc The number of arguments passed to ngIRCd on the command line.
 * @param argv An array containing all the arguments passed to ngIRCd.
 * @return Global exit code of ngIRCd, zero on success.
 */
GLOBAL int
main( int argc, const char *argv[] )
{
	bool ok, configtest = false;
	bool NGIRCd_NoDaemon = false;
	int i;
	size_t n;

	umask( 0077 );

	NGIRCd_SignalQuit = NGIRCd_SignalRestart = NGIRCd_SignalRehash = false;
	NGIRCd_Passive = false;
#ifdef DEBUG
	NGIRCd_Debug = false;
#endif
#ifdef SNIFFER
	NGIRCd_Sniffer = false;
#endif
	strlcpy( NGIRCd_ConfFile, SYSCONFDIR, sizeof( NGIRCd_ConfFile ));
	strlcat( NGIRCd_ConfFile, CONFIG_FILE, sizeof( NGIRCd_ConfFile ));

	Fill_Version( );

	/* Kommandozeile parsen */
	for( i = 1; i < argc; i++ )
	{
		ok = false;
		if(( argv[i][0] == '-' ) && ( argv[i][1] == '-' ))
		{
			/* Lange Option */

			if( strcmp( argv[i], "--config" ) == 0 )
			{
				if( i + 1 < argc )
				{
					/* Ok, there's an parameter left */
					strlcpy( NGIRCd_ConfFile, argv[i + 1], sizeof( NGIRCd_ConfFile ));

					/* next parameter */
					i++; ok = true;
				}
			}
			if( strcmp( argv[i], "--configtest" ) == 0 )
			{
				configtest = true;
				ok = true;
			}
#ifdef DEBUG
			if( strcmp( argv[i], "--debug" ) == 0 )
			{
				NGIRCd_Debug = true;
				ok = true;
			}
#endif
			if( strcmp( argv[i], "--help" ) == 0 )
			{
				Show_Version( );
				puts( "" ); Show_Help( ); puts( "" );
				exit( 1 );
			}
			if( strcmp( argv[i], "--nodaemon" ) == 0 )
			{
				NGIRCd_NoDaemon = true;
				ok = true;
			}
			if( strcmp( argv[i], "--passive" ) == 0 )
			{
				NGIRCd_Passive = true;
				ok = true;
			}
#ifdef SNIFFER
			if( strcmp( argv[i], "--sniffer" ) == 0 )
			{
				NGIRCd_Sniffer = true;
				ok = true;
			}
#endif
			if( strcmp( argv[i], "--version" ) == 0 )
			{
				Show_Version( );
				exit( 1 );
			}
		}
		else if(( argv[i][0] == '-' ) && ( argv[i][1] != '-' ))
		{
			/* Kurze Option */
			for( n = 1; n < strlen( argv[i] ); n++ )
			{
				ok = false;
#ifdef DEBUG
				if( argv[i][n] == 'd' )
				{
					NGIRCd_Debug = true;
					ok = true;
				}
#endif
				if( argv[i][n] == 'f' )
				{
					if(( ! argv[i][n + 1] ) && ( i + 1 < argc ))
					{
						/* Ok, next character is a blank */
						strlcpy( NGIRCd_ConfFile, argv[i + 1], sizeof( NGIRCd_ConfFile ));

						/* go to the following parameter */
						i++;
						n = strlen( argv[i] );
						ok = true;
					}
				}
				if( argv[i][n] == 'n' )
				{
					NGIRCd_NoDaemon = true;
					ok = true;
				}
				if( argv[i][n] == 'p' )
				{
					NGIRCd_Passive = true;
					ok = true;
				}
#ifdef SNIFFER
				if( argv[i][n] == 's' )
				{
					NGIRCd_Sniffer = true;
					ok = true;
				}
#endif
				if( argv[i][n] == 't' )
				{
					configtest = true;
					ok = true;
				}

				if( ! ok )
				{
					printf( "%s: invalid option \"-%c\"!\n", PACKAGE_NAME, argv[i][n] );
					printf( "Try \"%s --help\" for more information.\n", PACKAGE_NAME );
					exit( 1 );
				}
			}

		}
		if( ! ok )
		{
			printf( "%s: invalid option \"%s\"!\n", PACKAGE_NAME, argv[i] );
			printf( "Try \"%s --help\" for more information.\n", PACKAGE_NAME );
			exit( 1 );
		}
	}

	/* Debug-Level (fuer IRC-Befehl "VERSION") ermitteln */
	NGIRCd_DebugLevel[0] = '\0';
#ifdef DEBUG
	if( NGIRCd_Debug ) strcpy( NGIRCd_DebugLevel, "1" );
#endif
#ifdef SNIFFER
	if( NGIRCd_Sniffer )
	{
		NGIRCd_Debug = true;
		strcpy( NGIRCd_DebugLevel, "2" );
	}
#endif

	/* Soll nur die Konfigurations ueberprueft und ausgegeben werden? */
	if( configtest )
	{
		Show_Version( ); puts( "" );
		exit( Conf_Test( ));
	}
	
	while( ! NGIRCd_SignalQuit )
	{
		/* Initialize global variables */
		NGIRCd_Start = time( NULL );
		(void)strftime( NGIRCd_StartStr, 64, "%a %b %d %Y at %H:%M:%S (%Z)", localtime( &NGIRCd_Start ));

		NGIRCd_SignalRehash = false;
		NGIRCd_SignalRestart = false;
		NGIRCd_SignalQuit = false;

		/* Initialize modules, part I */
		Log_Init( ! NGIRCd_NoDaemon );
		Conf_Init( );

		if (!NGIRCd_Init( NGIRCd_NoDaemon )) {
			Log(LOG_WARNING, "Fatal: Initialization failed");
			exit(1);
		}

		/* Initialize modules, part II: these functions are eventually
		 * called with already dropped privileges ... */
		Lists_Init( );
		Channel_Init( );
		Client_Init( );
#ifdef ZEROCONF
		Rendezvous_Init( );
#endif
		Conn_Init( );

#ifdef DEBUG
		/* Redirect stderr handle to "error file" for debugging
		 * when not running in "no daemon" mode: */
		if( ! NGIRCd_NoDaemon ) Log_InitErrorfile( );
#endif

		/* Signal-Handler initialisieren */
		Initialize_Signal_Handler( );

		/* Protokoll- und Server-Identifikation erzeugen. Die vom ngIRCd
		 * beim PASS-Befehl verwendete Syntax sowie die erweiterten Flags
		 * sind in doc/Protocol.txt beschrieben. */
#ifdef IRCPLUS
		snprintf( NGIRCd_ProtoID, sizeof NGIRCd_ProtoID, "%s%s %s|%s:%s", PROTOVER, PROTOIRCPLUS, PACKAGE_NAME, PACKAGE_VERSION, IRCPLUSFLAGS );
#ifdef ZLIB
		strcat( NGIRCd_ProtoID, "Z" );
#endif
		if( Conf_OperCanMode ) strcat( NGIRCd_ProtoID, "o" );
#else
		snprintf( NGIRCd_ProtoID, sizeof NGIRCd_ProtoID, "%s%s %s|%s", PROTOVER, PROTOIRC, PACKAGE_NAME, PACKAGE_VERSION );
#endif
		strlcat( NGIRCd_ProtoID, " P", sizeof NGIRCd_ProtoID );
#ifdef ZLIB
		strlcat( NGIRCd_ProtoID, "Z", sizeof NGIRCd_ProtoID );
#endif
		Log( LOG_DEBUG, "Protocol and server ID is \"%s\".", NGIRCd_ProtoID );

		/* Vordefinierte Channels anlegen */
		Channel_InitPredefined( );

		/* Listen-Ports initialisieren */
		if( Conn_InitListeners( ) < 1 )
		{
			Log( LOG_ALERT, "Server isn't listening on a single port!" );
			Log( LOG_ALERT, "%s exiting due to fatal errors!", PACKAGE_NAME );
			Pidfile_Delete( );
			exit( 1 );
		}
		
		/* Hauptschleife */
		Conn_Handler( );

		/* Alles abmelden */
		Conn_Exit( );
#ifdef ZEROCONF
		Rendezvous_Exit( );
#endif
		Client_Exit( );
		Channel_Exit( );
		Lists_Exit( );
		Log_Exit( );
	}
	Pidfile_Delete( );

	return 0;
} /* main */


/**
 * Generate ngIRCd "version string".
 * This string is generated once and then stored in NGIRCd_Version for
 * further usage, for example by the IRC command VERSION and the --version
 * command line switch.
 */
static void
Fill_Version( void )
{
	NGIRCd_VersionAddition[0] = '\0';

#ifdef SYSLOG
	strlcpy( NGIRCd_VersionAddition, "SYSLOG", sizeof NGIRCd_VersionAddition );
#endif
#ifdef ZLIB
	if( NGIRCd_VersionAddition[0] )
		strlcat( NGIRCd_VersionAddition, "+", sizeof NGIRCd_VersionAddition );

	strlcat( NGIRCd_VersionAddition, "ZLIB", sizeof NGIRCd_VersionAddition );
#endif
#ifdef TCPWRAP
	if( NGIRCd_VersionAddition[0] )
			strlcat( NGIRCd_VersionAddition, "+", sizeof NGIRCd_VersionAddition );

	strlcat( NGIRCd_VersionAddition, "TCPWRAP", sizeof NGIRCd_VersionAddition );
#endif
#ifdef ZEROCONF
	if( NGIRCd_VersionAddition[0] )
		strlcat( NGIRCd_VersionAddition, "+", sizeof NGIRCd_VersionAddition );

	strlcat( NGIRCd_VersionAddition, "ZEROCONF", sizeof NGIRCd_VersionAddition );
#endif
#ifdef IDENTAUTH
	if( NGIRCd_VersionAddition[0] )
		strlcat( NGIRCd_VersionAddition, "+", sizeof NGIRCd_VersionAddition );

	strlcat( NGIRCd_VersionAddition, "IDENT", sizeof NGIRCd_VersionAddition );
#endif
#ifdef DEBUG
	if( NGIRCd_VersionAddition[0] )
		strlcat( NGIRCd_VersionAddition, "+", sizeof NGIRCd_VersionAddition );

	strlcat( NGIRCd_VersionAddition, "DEBUG", sizeof NGIRCd_VersionAddition );
#endif
#ifdef SNIFFER
	if( NGIRCd_VersionAddition[0] )
		strlcat( NGIRCd_VersionAddition, "+", sizeof NGIRCd_VersionAddition );

	strlcat( NGIRCd_VersionAddition, "SNIFFER", sizeof NGIRCd_VersionAddition );
#endif
#ifdef STRICT_RFC
	if( NGIRCd_VersionAddition[0] )
		strlcat( NGIRCd_VersionAddition, "+", sizeof NGIRCd_VersionAddition );

	strlcat( NGIRCd_VersionAddition, "RFC", sizeof NGIRCd_VersionAddition );
#endif
#ifdef IRCPLUS
	if( NGIRCd_VersionAddition[0] )
		strlcat( NGIRCd_VersionAddition, "+", sizeof NGIRCd_VersionAddition );

	strlcat( NGIRCd_VersionAddition, "IRCPLUS", sizeof NGIRCd_VersionAddition );
#endif

	if( NGIRCd_VersionAddition[0] )
		strlcat( NGIRCd_VersionAddition, "-", sizeof( NGIRCd_VersionAddition ));

	strlcat( NGIRCd_VersionAddition, TARGET_CPU, sizeof( NGIRCd_VersionAddition ));
	strlcat( NGIRCd_VersionAddition, "/", sizeof( NGIRCd_VersionAddition ));
	strlcat( NGIRCd_VersionAddition, TARGET_VENDOR, sizeof( NGIRCd_VersionAddition ));
	strlcat( NGIRCd_VersionAddition, "/", sizeof( NGIRCd_VersionAddition ));
	strlcat( NGIRCd_VersionAddition, TARGET_OS, sizeof( NGIRCd_VersionAddition ));

#ifdef CVSDATE
	snprintf( NGIRCd_Version, sizeof NGIRCd_Version,"%s %s(%s)-%s", PACKAGE_NAME, PACKAGE_VERSION, CVSDATE, NGIRCd_VersionAddition);
#else
	snprintf( NGIRCd_Version, sizeof NGIRCd_Version, "%s %s-%s", PACKAGE_NAME, PACKAGE_VERSION, NGIRCd_VersionAddition);
#endif
} /* Fill_Version */


/**
 * Reload the server configuration file.
 */
GLOBAL void
NGIRCd_Rehash( void )
{
	char old_name[CLIENT_ID_LEN];

	Log( LOG_NOTICE|LOG_snotice, "Re-reading configuration NOW!" );
	NGIRCd_SignalRehash = false;

	/* Close down all listening sockets */
	Conn_ExitListeners( );

	/* Remember old server name */
	strlcpy( old_name, Conf_ServerName, sizeof old_name );

	/* Re-read configuration ... */
	Conf_Rehash( );

	/* Recover old server name: it can't be changed during run-time */
	if( strcmp( old_name, Conf_ServerName ) != 0 )
	{
		strlcpy( Conf_ServerName, old_name, sizeof Conf_ServerName );
		Log( LOG_ERR, "Can't change \"ServerName\" on runtime! Ignored new name." );
	}

	/* Create new pre-defined channels */
	Channel_InitPredefined( );
	
	/* Start listening on sockets */
	Conn_InitListeners( );

	/* Sync configuration with established connections */
	Conn_SyncServerStruct( );

	Log( LOG_NOTICE|LOG_snotice, "Re-reading of configuration done." );
} /* NGIRCd_Rehash */


/**
 * Initialize the signal handler.
 */
static void
Initialize_Signal_Handler( void )
{
	/* Signal-Handler initialisieren: einige Signale
	 * werden ignoriert, andere speziell behandelt. */

#ifdef HAVE_SIGACTION
	/* sigaction() ist vorhanden */

	struct sigaction saction;

	/* Signal-Struktur initialisieren */
	memset( &saction, 0, sizeof( saction ));
	saction.sa_handler = Signal_Handler;
#ifdef SA_RESTART
	saction.sa_flags |= SA_RESTART;
#endif
#ifdef SA_NOCLDWAIT
	saction.sa_flags |= SA_NOCLDWAIT;
#endif

	/* Signal-Handler einhaengen */
	sigaction( SIGINT, &saction, NULL );
	sigaction( SIGQUIT, &saction, NULL );
	sigaction( SIGTERM, &saction, NULL);
	sigaction( SIGHUP, &saction, NULL);
	sigaction( SIGCHLD, &saction, NULL);

	/* einige Signale ignorieren */
	saction.sa_handler = SIG_IGN;
	sigaction( SIGPIPE, &saction, NULL );
#else
	/* kein sigaction() vorhanden */

	/* Signal-Handler einhaengen */
	signal( SIGINT, Signal_Handler );
	signal( SIGQUIT, Signal_Handler );
	signal( SIGTERM, Signal_Handler );
	signal( SIGHUP, Signal_Handler );
	signal( SIGCHLD, Signal_Handler );

	/* einige Signale ignorieren */
	signal( SIGPIPE, SIG_IGN );
#endif
} /* Initialize_Signal_Handler */


/**
 * Signal handler of ngIRCd.
 * This function is called whenever ngIRCd catches a signal sent by the
 * user and/or the system to it. For example SIGTERM and SIGHUP.
 * @param Signal Number of the signal to handle.
 */
static void
Signal_Handler( int Signal )
{
	switch( Signal )
	{
		case SIGTERM:
		case SIGINT:
		case SIGQUIT:
			/* wir soll(t)en uns wohl beenden ... */
			NGIRCd_SignalQuit = true;
			break;
		case SIGHUP:
			/* Konfiguration neu einlesen: */
			NGIRCd_SignalRehash = true;
			break;
		case SIGCHLD:
			/* Child-Prozess wurde beendet. Zombies vermeiden: */
			while( waitpid( -1, NULL, WNOHANG ) > 0);
			break;
#ifdef DEBUG
		default:
			/* unbekanntes bzw. unbehandeltes Signal */
			Log( LOG_DEBUG, "Got signal %d! Ignored.", Signal );
#endif
	}
} /* Signal_Handler */


/**
 * Display copyright and version information of ngIRCd on the console.
 */
static void
Show_Version( void )
{
	puts( NGIRCd_Version );
	puts( "Copyright (c)2001-2005 by Alexander Barton (<alex@barton.de>)." );
	puts( "Homepage: <http://arthur.ath.cx/~alex/ngircd/>\n" );
	puts( "This is free software; see the source for copying conditions. There is NO" );
	puts( "warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE." );
} /* Show_Version */


/**
 * Display a short help text on the console.
 * This help depends on the configuration of the executable and only shows
 * options that are actually enabled.
 */
static void
Show_Help( void )
{
#ifdef DEBUG
	puts( "  -d, --debug        log extra debug messages" );
#endif
	puts( "  -f, --config <f>   use file <f> as configuration file" );
	puts( "  -n, --nodaemon     don't fork and don't detach from controlling terminal" );
	puts( "  -p, --passive      disable automatic connections to other servers" );
#ifdef SNIFFER
	puts( "  -s, --sniffer      enable network sniffer and display all IRC traffic" );
#endif
	puts( "  -t, --configtest   read, validate and display configuration; then exit" );
 	puts( "      --version      output version information and exit" );
	puts( "      --help         display this help and exit" );
} /* Show_Help */


/**
 * Delete the file containing the process ID (PID).
 */
static void
Pidfile_Delete( void )
{
	/* Pidfile configured? */
	if( ! Conf_PidFile[0] ) return;

#ifdef DEBUG
	Log( LOG_DEBUG, "Removing PID file (%s) ...", Conf_PidFile );
#endif

	if( unlink( Conf_PidFile ))
		Log( LOG_ERR, "Error unlinking PID file (%s): %s", Conf_PidFile, strerror( errno ));
} /* Pidfile_Delete */


/**
 * Create the file containing the process ID of ngIRCd ("PID file").
 * @param pid The process ID to be stored in this file.
 */
static void
Pidfile_Create( long pid )
{
	int pidfd;
	char pidbuf[64];
	int len;

	/* Pidfile configured? */
	if( ! Conf_PidFile[0] ) return;

#ifdef DEBUG
	Log( LOG_DEBUG, "Creating PID file (%s) ...", Conf_PidFile );
#endif

	pidfd = open( Conf_PidFile, O_RDWR|O_CREAT|O_EXCL, S_IRUSR|S_IWUSR|S_IRGRP|S_IROTH);
	if ( pidfd < 0 ) {
		Log( LOG_ERR, "Error writing PID file (%s): %s", Conf_PidFile, strerror( errno ));
		return;
	}

	len = snprintf( pidbuf, sizeof pidbuf, "%ld\n", pid );
	if (len < 0|| len < (int)sizeof pid) {
		Log( LOG_ERR, "Error converting pid");
		return;
	}
	
	if( write( pidfd, pidbuf, len) != len)
		Log( LOG_ERR, "Can't write PID file (%s): %s", Conf_PidFile, strerror( errno ));

	if( close(pidfd) != 0 )
		Log( LOG_ERR, "Error closing PID file (%s): %s", Conf_PidFile, strerror( errno ));
} /* Pidfile_Create */


/**
 * Redirect stdin, stdout and stderr to apropriate file handles.
 */
static void
Setup_FDStreams( void )
{
	int fd;

	/* Test if we can open /dev/null for reading and writing. If not
	 * we are most probably chrooted already and the server has been
	 * restarted. So we simply don't try to redirect stdXXX ... */
	fd = open( "/dev/null", O_RDWR );
	if ( fd < 0 ) {
		Log(LOG_WARNING, "Could not open /dev/null: %s", strerror(errno));	
		return;
	} 

	fflush(stdout);
	fflush(stderr);

	/* Create new stdin(0), stdout(1) and stderr(2) descriptors */
	dup2( fd, 0 ); dup2( fd, 1 ); dup2( fd, 2 );

	/* Close newly opened file descriptor if not stdin/out/err */
	if( fd > 2 ) close( fd );
} /* Setup_FDStreams */


static bool
NGIRCd_getNobodyID(unsigned int *uid, unsigned int *gid )
{
	struct passwd *pwd;

	pwd = getpwnam("nobody");
	if (!pwd) return false;

	if ( !pwd->pw_uid || !pwd->pw_gid)
		return false;

	*uid = pwd->pw_uid;	
	*gid = pwd->pw_gid;
	endpwent();

	return true;	
}


static bool
NGIRCd_Init( bool NGIRCd_NoDaemon ) 
{
	static bool initialized;
	bool chrooted = false;
	struct passwd *pwd;
	struct group *grp;
	int real_errno;
	long pid;

	if (initialized)
		return true;

	if( Conf_Chroot[0] ) {
		if( chdir( Conf_Chroot ) != 0 ) {
			Log( LOG_ERR, "Can't chdir() in ChrootDir (%s): %s", Conf_Chroot, strerror( errno ));
			return false;
		}

		if( chroot( Conf_Chroot ) != 0 ) {
			if (errno != EPERM) {
				Log( LOG_ERR, "Can't change root directory to \"%s\": %s",
								Conf_Chroot, strerror( errno ));

				return false;
			}
		} else {
			chrooted = true;
			Log( LOG_INFO, "Changed root and working directory to \"%s\".", Conf_Chroot );
		}
	}

	if (Conf_UID == 0) {
		Log(LOG_INFO, "ServerUID must not be 0, using \"nobody\" instead.", Conf_UID);

  		if (! NGIRCd_getNobodyID(&Conf_UID, &Conf_GID)) {
			Log(LOG_WARNING, "Could not get user/group ID of user \"nobody\": %s",
					errno ? strerror(errno) : "not found" );
			return false;
		}
	}

	if (getgid() != Conf_GID) {
		/* Change group ID */
		if (setgid(Conf_GID) != 0) {
			real_errno = errno;
			Log( LOG_ERR, "Can't change group ID to %u: %s", Conf_GID, strerror( errno ));
			if (real_errno != EPERM) 
				return false;
		}
	}

	if (getuid() != Conf_UID) {
		/* Change user ID */
		if (setuid(Conf_UID) != 0) {
			real_errno = errno;
			Log(LOG_ERR, "Can't change user ID to %u: %s", Conf_UID, strerror(errno));
			if (real_errno != EPERM) 
				return false;
		}
	}

	initialized = true;

	/* Normally a child process is forked which isn't any longer
	 * connected to ther controlling terminal. Use "--nodaemon"
	 * to disable this "daemon mode" (useful for debugging). */
	if ( ! NGIRCd_NoDaemon ) {
		pid = (long)fork( );
		if( pid > 0 ) {
			/* "Old" process: exit. */
			exit( 0 );
		}
		if( pid < 0 ) {
			/* Error!? */
			fprintf( stderr, "%s: Can't fork: %s!\nFatal error, exiting now ...\n",
								PACKAGE_NAME, strerror( errno ));
			exit( 1 );
		}

		/* New child process */
		(void)setsid( );
		chdir( "/" );

		/* Detach stdin, stdout and stderr */
		Setup_FDStreams( );
	}
	pid = getpid();

	Pidfile_Create( pid );

	/* Check UID/GID we are running as, can be different from values
	 * configured (e. g. if we were already started with a UID>0. */
	Conf_UID = getuid();
	Conf_GID = getgid();

	pwd = getpwuid( Conf_UID );
	grp = getgrgid( Conf_GID );

	Log( LOG_INFO, "Running as user %s(%ld), group %s(%ld), with PID %ld.",
				pwd ? pwd->pw_name : "unknown", Conf_UID,
				grp ? grp->gr_name : "unknown", Conf_GID, pid);

	if ( chrooted ) {
		Log( LOG_INFO, "Running chrooted, chrootdir \"%s\".",  Conf_Chroot );
		return true;
	} else {
		Log( LOG_INFO, "Not running chrooted." );
	}

	/* Change working directory to home directory of the user
	 * we are running as (only when running in daemon mode and not in chroot) */
	
	if ( pwd ) {
		if (!NGIRCd_NoDaemon ) {
			if( chdir( pwd->pw_dir ) == 0 ) 
				Log( LOG_DEBUG, "Changed working directory to \"%s\" ...", pwd->pw_dir );
			else 
				Log( LOG_ERR, "Can't change working directory to \"%s\": %s",
								pwd->pw_dir, strerror( errno ));
		}
	} else {
		Log( LOG_ERR, "Can't get user informaton for UID %d!?", Conf_UID );
	}

return true;
}

/* -eof- */
